custom/icds/location_reassignment/processor.py

```diff
@@ -1,29 +1,79 @@
+from django.db import transaction
 from django.utils.functional import cached_property
 
 from corehq.apps.locations.models import LocationType, SQLLocation
 from custom.icds.location_reassignment.const import SPLIT_OPERATION
-from custom.icds.location_reassignment.exceptions import InvalidTransitionError
+from custom.icds.location_reassignment.exceptions import (
+    InvalidTransitionError,
+    LocationCreateError,
+)
 from custom.icds.location_reassignment.utils import deprecate_locations
 
 
 class Processor(object):
-    def __init__(self, domain, transitions, site_codes):
+    def __init__(self, domain, transitions, new_location_details, site_codes):
         """
         :param domain: domain
         :param transitions: transitions in format generated by Parser
+        :param new_location_details: details necessary to create new locations
         :param site_codes: site codes of all locations undergoing transitions
         """
         self.domain = domain
-        self.location_type_codes = list(map(lambda lt: lt.code, LocationType.objects.by_domain(self.domain)))
+        self.location_types_by_code = {lt.code: lt for lt in LocationType.objects.by_domain(self.domain)}
         self.transitions = transitions
+        self.new_location_details = new_location_details
         self.site_codes = site_codes
 
     def process(self):
+        self._create_new_locations()
         # process each sheet, in reverse order of hierarchy
-        for location_type_code in list(reversed(self.location_type_codes)):
+        for location_type_code in list(reversed(list(self.location_types_by_code.keys()))):
             for operation, transitions in self.transitions[location_type_code].items():
                 self._perform_transitions(operation, transitions)
 
+    def _create_new_locations(self):
+        locations_by_site_code = self._get_existing_parent_locations()
+
+        with transaction.atomic():
+            for location_type_code in self.location_types_by_code:
+                new_locations_details = self.new_location_details.get(location_type_code, [])
+                for details in new_locations_details:
+                    parent_location = None
+                    if details['parent_site_code']:
+                        parent_location = locations_by_site_code[details['parent_site_code']]
+                    location = SQLLocation.objects.create(
+                        domain=self.domain, site_code=details['site_code'], name=details['name'],
+                        parent=parent_location,
```

Comment:
Couldn't you use `parent_id=details['parent_site_code']`? I'd think that would throw an error if the parent location didn't exist, and you could catch that here instead of doing the lookups and error checking yourself.

Reply:
that would need to be `parent_id=parent_location.id`. Currently the `id` is not available, so we would need pass the parent location id from parser itself.
My reason for error checking earlier is to 
1. catch a missing location in advance and not even begin creating locations and then rollback in case of failure.
2. load locations in bulk
that sound reasonable?

Reply:
Ah, I was confused between site code and id. This makes sense then.